substrate/frame/identity/src/lib.rs

```diff
@@ -1164,107 +1231,178 @@ pub mod pallet {
 			// ensure `username` maps to `origin` (i.e. has already been set by an authority).
 			let who = ensure_signed(origin)?;
 			let account_of_username =
-				AccountOfUsername::<T>::get(&username).ok_or(Error::<T>::NoUsername)?;
+				UsernameInfoOf::<T>::get(&username).ok_or(Error::<T>::NoUsername)?.owner;
 			ensure!(who == account_of_username, Error::<T>::InvalidUsername);
-			let (registration, _maybe_username) =
-				IdentityOf::<T>::get(&who).ok_or(Error::<T>::NoIdentity)?;
-			IdentityOf::<T>::insert(&who, (registration, Some(username.clone())));
+			UsernameOf::<T>::insert(&who, username.clone());
 			Self::deposit_event(Event::PrimaryUsernameSet { who: who.clone(), username });
 			Ok(())
 		}
 
-		/// Remove a username that corresponds to an account with no identity. Exists when a user
-		/// gets a username but then calls `clear_identity`.
+		/// Remove a username that corresponds to an account with no primary username. Exists when a
+		/// user gets a username but then the authority removes it.
 		#[pallet::call_index(21)]
 		#[pallet::weight(T::WeightInfo::remove_dangling_username())]
 		pub fn remove_dangling_username(
```

Comment:
I'm a bit surprised by this function.

In general a user can't remove his usernames, only authorities and force origin can.
But if force origin kill a primary username of someone, then this user can remove all his other usernames?

I would expect to only be able to set a new primary (which he is able to).

What is the reason of this sudden increase in privileges?

Reply:
It's not ideal but the reasoning was to have a function that can do "cleanup" work in a permissionless way. If `ForceOrigin` wants to punish an account, they'd presumably look up their primary username and kill that, and next other users can remove any other usernames. The problem is that, in practice, `ForceOrigin` is governance, which means there's a delay between when we make it public that a username will be removed and the actual `kill_username` call taking place, so the user would technically know in advance and prepare by setting another username as primary immediately after the existing one is removed, therefore restarting the whole process.

One fix would be to disallow setting a primary username if you currently don't have any (which would happen only if it was removed, either by governance or by an authority), but this just feels like a hack because it relies on this loose assumption.

The other option would be to remove this call entirely and let governance delete whatever it wants with the referendum proposer doing the username "mining" work offchain. I think I'd go with this one.

Reply:
I see, I'm fine with anything.